functions/api/roles.js

```diff
@@ -0,0 +1,149 @@
+    const doc = await rolesRef.get()
+    if (doc.exists) {
+        const err = new Error(`Role ${id} already exists. Please update if needed.`)
+        err.statusCode = 404
+        next(err)
+        return
+    }
+    await db.collection(constants.ROLES).doc(id).set({
+        ...req.body,
+        createdDate: new Date(),
+        lastUpdatedDate: new Date(),
+        label: id.toUpperCase()
+    })
+    res.status(201).send({ 'message': 'created successfully' })
+})
+
+/*
+1.Returns all roles in roles collection along with permissions, description, status.
+2.Admin and System are the roles supported currently.
+*/
+router.get('/', async (req, res, next) => {
+    let rolesRef = db.collection(constants.ROLES)
+    const snapshot = await rolesRef.get()
+    const allRoles = []
+    snapshot.forEach(doc => {
+        const roles = formatDate(doc.data())
+        allRoles.push({ id: doc.id, ...roles })
+    })
+    const response = {
+        roles: allRoles,
+        totalRoles: allRoles.length
+    }
+    res.status(200).send(response)
+})
+
+/*
+1.Returns a specific role and its permissions, description, status
+*/
+router.get('/:role', async (req, res, next) => {
+    let rolesRef = db.collection(constants.ROLES).doc(req.params.role)
+    const doc = await rolesRef.get()
+    if (!doc.exists) {
+        const error = new Error(`${req.params.role} not found in firestore`)
+        error.statusCode = 404
+        next(error)
+        return
+    } else {
+        const role = formatDate(doc.data())
+        res.status(200).send({ id: doc.id, ...role })
+    }
+})
+
+/*
+1.Updates the role record with given permissions
+2.Status of the role, description of role can also be updated
+*/
+router.put('/', async (req, res, next) => {
+    if(req.body.createdDate){
+        delete req.body.createdDate
+    }
+    const { error } = validateInput(req.body, constants.UPDATE)
+    if (error) {
+        const err = new Error(error.details[0].message)
+        err.statusCode = 400
+        next(err)
+        return
+    }
+    const { id } = req.body
+    delete req.body.id
+    let rolesRef = db.collection(constants.ROLES).doc(id)
+    const doc = await rolesRef.get()
+    if (!doc.exists) {
+        const error = new Error(`Requested ${id} is not present in firestore`)
+        error.statusCode = 404
+        next(error)
+        return
+    }
+    await rolesRef.update({
+        ...req.body,
+        lastUpdatedDate: new Date()
+    })
+    res.sendStatus(204)
+
+})
+
+/*
+1.Deletes the role record
+*/
+router.delete('/:role', async (req, res, next) => {
+    let rolesRef = db.collection(constants.ROLES).doc(req.params.role)
+    const doc = await rolesRef.get()
+    if (!doc.exists) {
+        const error = new Error(`${req.params.role} not found in firestore`)
+        error.statusCode = 404
+        next(error)
+        return
+    }
+    await db.collection(constants.ROLES).doc(req.params.role).delete()
```

Comment:
Small observation: you already have rolesRef which can be reused here.

Reply:
oops, may be i missed it. Took care in users.Will correct in my next PR